### PR TITLE
Profpoint trap event collector

### PR DESCRIPTION
**Pull request by daryakoval**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-3170

## Description
Event collector 

## Minimum version of Cortex XSOAR
- [x] 6.8.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+Use the Proofpoint Threat Response integration to orchestrate and automate incident response.
+
+## Configure Proofpoint Threat Response Event Collector on Cortex XSOAR
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/README.md:3]:
```suggestion
## Configure Proofpoint Threat Response Event Collector on Cortex XSIAM
```

File: Packs/ProofpointThreatResponse/ReleaseNotes/1_0_13.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+#### Integrations
+##### New: Proofpoint Threat Response Event Collector
+Use the Proofpoint Threat Response integration to orchestrate and automate incident response. (Available from Cortex XSOAR 6.8.0).
+
+#### Modeling Rules
+##### New: Proofpoint Threat Response Modeling Rule
+Added the *Email* modeling rule. (Available from Cortex XSOAR 6.8.0).
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/ReleaseNotes/1_0_13.md:8]:
```suggestion
Use the Proofpoint Threat Response integration to orchestrate and automate incident response. (Available from Cortex XSIAM 3.3.0).

#### Modeling Rules
##### New: Proofpoint Threat Response Modeling Rule
Added the *Email* modeling rule. (Available from Cortex XSIAM 3.3.0).
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector_description.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+In order to create the API Key go to:
+
+  Settings -> API Key -> Add api key
+  
+  Note: This is a beta Integration, which lets you implement and test pre-release software. Since the integration is beta, it might contain bugs. Updates to the integration during the beta phase might include non-backward compatible features. We appreciate your feedback on the quality and usability of the integration to help us identify issues, fix them, and continually improve.
+
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector_description.md:6]:
```suggestion  
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector_description.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+In order to create the API Key go to:
+
+  Settings -> API Key -> Add api key
+  
+  Note: This is a beta Integration, which lets you implement and test pre-release software. Since the integration is beta, it might contain bugs. Updates to the integration during the beta phase might include non-backward compatible features. We appreciate your feedback on the quality and usability of the integration to help us identify issues, fix them, and continually improve.
+  
+- The ***proofpoint-tr-ingest-alert*** command requires some JSON format arguments. Please make sure the inputed JSONs have uniqe characters (such as `"`) escaped before entering them (e.g. `\"`).
+- For more information regarding the JSON objects, please see Proofpoint TRAP documentation under "JSON Alert Source 2.0".
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector_description.md:8]:
```suggestion
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+    demisto.debug(f"End of current fetch function with last_fetch {str(last_fetch)} and last_fetched_id"
+                  f" {str(last_fetched_id)}")
+
+    last_run = {
+        'last_fetch': last_fetch,
+        'last_fetched_incident_id': last_fetched_id
+    }
+
+    demisto.info(f'extracted {len(incidents)} events')
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py:308]:
```suggestion
    demisto.debug(f'Fetched {len(incidents)} events')
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        demisto.debug(f"End of the current batch loop with {str(len(incidents_list))} events")
+
+    # fetching the last batch when created_before is bigger then current time = fetching new events
+    if len(incidents_list) < fetch_limit:   # type: ignore[operator]
+        # fetching the last batch
+        request_params['created_before'] = current_time.isoformat().split('.')[0] + 'Z'
+        new_incidents = get_new_incidents(client, request_params, last_fetched_id)
+        incidents_list.extend(new_incidents)
+
+        demisto.info(
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py:260]:
```suggestion
        demisto.debug(
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    last_fetch = last_run.get('last_fetch', {})
+    last_fetched_id = last_run.get('last_fetched_incident_id', {})
+
+    for state in incidents_states:
+        if not last_fetch.get(state):
+            last_fetch[state] = first_fetch
+        if not last_fetched_id.get(state):
+            last_fetched_id[state] = '0'
+
+    incidents = []
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py:281]:
```suggestion
    events = []
```

**Review by dantavori — CHANGES_REQUESTED** [review]

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+requests.packages.urllib3.disable_warnings()  # pylint: disable=no-member
+
+TIME_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+PRODUCT = 'threat_response'
+VENDOR = 'proofpoint'
+
+
+class Client(BaseClient):
+    """
+    Client will implement the service API, and should not contain any Demisto logic.
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py:14]:
```suggestion
    Client will implement the service API, and should not contain any logic.
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  hiddenusername: true
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- defaultvalue: 'false'
+  display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml:24]: please move to bottom of the params

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  - ignored
+  required: false
+  type: 16
+- additionalinfo: You can find this value by navigating to Sources -> JSON event source -> POST URL.
+  display: POST URL of the JSON alert source.
+  hidden: false
+  name: post_url_id
+  required: false
+  type: 0
+description: Use the Proofpoint Threat Response integration to orchestrate and automate incident response.
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml:67]: please verify this description with a TPM

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector_description.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+In order to create the API Key go to:
+
+  Settings -> API Key -> Add api key
+  
+- If the **fetch-incidents** result in timeout, please consider changing **fetch-delta** and **fetch-limit** parameters 
+which will limit the api calls and the result.
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector_description.md:6]:
```suggestion
**Note:** if the **fetch-events** command results in a timeout, consider changing **fetch_delta** and **fetch_limit** parameters to reduce the number of API calls and results.
```

File: Packs/ProofpointThreatResponse/ModelingRules/ProofpointThreatResponseModelingRules/ProofpointThreatResponseModelingRules.yml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+from_xdr_version: 3.4.0
+fromversion: 6.8.0
+name: Proofpoint Threat Response Modeling Rule
+id: proofpoint_threat_response_modeling_rule
+tags: proofpoint
+rules: ''
+schema: ''
+to_xdr_version: 3.4.9
+toversion: 6.9.9
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/ModelingRules/ProofpointThreatResponseModelingRules/ProofpointThreatResponseModelingRules.yml:9]:
```suggestion
```
as Shahaf said last week, if the issues with the new modeling rules are not resolved until you are ready to merge, let's remove them and we'll add them later

File: Packs/ProofpointThreatResponse/pack_metadata.json [hunk 13]
```diff
@@ -1,8 +1,8 @@
 {
-    "name": "Proofpoint Threat Response (Beta)",
+    "name": "Proofpoint Threat Response",
     "description": "Use the Proofpoint Threat Response integration to orchestrate and automate incident response.",
     "support": "xsoar",
-    "currentVersion": "1.0.13",
+    "currentVersion": "1.0.14",
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/pack_metadata.json:5]: should be major update for new integration - 2.0.0

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml [hunk 14]
```diff
@@ -0,0 +1,153 @@
+category: Email Gateway
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml:1]: shouldn't it be `Analytics & SIEM`?

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+
+def main():  # pragma: no cover
+    """main function, parses params and runs command functions
+        """
+    args = demisto.args()
+    command = demisto.command()
+    params = demisto.params()
+
+    api_key = params.get('credentials', {}).get('password')
+    base_url = demisto.params().get('url')
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py:341]:
```suggestion
    base_url = params.get('url')
```

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+        'last_fetched_incident_id': last_fetched_id
+    }
+
+    demisto.debug(f'Fetched {len(incidents)} events')
+
+    events = get_events_from_incidents(incidents)
+    return events, last_run
+
+
+def get_events_from_incidents(incidents):
```
Comment (dantavori) [inline Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.py:317]: please add a docstring

**Review by dantavori — COMMENTED** [review]
looks great, please assign a doc reviewer
I will probably approve after the demo next week

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml (hunk 14 above)
Comment (daryakoval) [reply Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml:1]: No, it is the same as Integration.

File: Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml (hunk 10 above)
Comment (daryakoval) [reply Packs/ProofpointThreatResponse/Integrations/ProofpointThreatResponseEventCollector/ProofpointThreatResponseEventCollector.yml:67]: Meital is out of the office.  This is the description of the integration that fetches incidents.

**Comment by ShirleyDenkberg** [comment]
@daryakoval Shouldn't there be a pack README?
@dantavori Doc review completed.

**Review by dantavori — APPROVED** [review]

**Comment by daryakoval** [comment]
Force merge needed: Proofpoint pack was only for "xsoar" marketplace. I added event collector and added to the pack metadata "marketplacev2" too. 
The build fails on the "pre-update" step, since before the update there is no such pack in the marketplacev2, and after the update the "post-update" passes.

**Pull request merged**